### PR TITLE
Make plasticity great again

### PR DESCRIPTION
**Pull request by albert-de-montserrat**

### CODE REVIEW

**Comment by albert-de-montserrat** [comment]
@boriskaus @aelligp

**Comment by albert-de-montserrat** [comment]
And this is by doing the stress update on the cell centers, and afterwards calculating the shear stress on vertices by averaging the effective visco-elasto-plastic viscosity from centers to vertices.

Shear bands look much more localised, but so far I didn't manage to make it fully converge.

**Comment by albert-de-montserrat** [comment]
@boriskaus @aelligp

**Comment by luraess** [comment]
Important is that vertex-to-cell and ell-to-vertex is done in a consistent way in order to have a converging algorithm. 
Also, as in the README https://github.com/PTsolvers/Stokes2D_simpleVEP, there may be some different behaviour dependent of the formulation (cell vs vertex vs cell+vertex) and on how interpolation is performed...

**Comment by boriskaus** [comment]
The results look pretty good to me. Yet is this regularised in the sense that increasing resolution gives the same maximum strainrate (not much chance of convergence without that). 
As @luraess says, the interpolation matters a lot

**Comment by luraess** [comment]
Is there visco-plastic regularisation in there (Thibault's approach)?

**Comment by albert-de-montserrat** [comment]
Yes this is regularised plasticity, same as in Thibault's approach.

File: src/stokes/Stokes3D.jl [hunk 1]
```diff
@@ -346,15 +302,21 @@ function JustRelax.solve!(
             @parallel (@idx ni) compute_ρg!(ρg[3], rheology, args)
 
             ν = 1e-3
```
Comment (aelligp) [inline src/stokes/Stokes3D.jl:304]: is the damping factor a remnant here? as you assign 1.0 in the function itself. 
Same thing at L491.

Comment (albert-de-montserrat) [reply src/stokes/Stokes3D.jl:304]: 1.0 is for testing the plasticity on this particular benchmark, it will use 1e-2 or 1e-3 once merged

**Comment by albert-de-montserrat** [comment]
These are the stress-strain curves and number of PT iterations for Von Mises, Drucker-Prager, and regularised DruckerPrager with:

- **Solid lines**: stress update don't all in the cell centers (shear strain rate is linearly interpolated to cell center), then shear stress is linearly interpolated from cell centers to cell vertices.

- **Dashed lines**: stress update as before, but now effective visco-elasto-plastic viscosity is averaged at the vertices to compute the shear stress.

stress_curve

**Comment by boriskaus** [comment]
interesting - so the averaging of vep viscosity works best?

**Comment by albert-de-montserrat** [comment]
Requires a bit more of PT iterations when plasticity is localising, but looks better indeed.

**Comment by luraess** [comment]
What tells you dashed lines behaves best? It looks like both results and convergence history is better for plain lines compared to dashed ones. Would be interesting to test at higher res to draw a conclusion.

**Comment by albert-de-montserrat** [comment]
Some shear band plots looked nicer with dashed line approach. It will take a bit to run all the models, but you are right, at least for regularised DP the shear band looks nicer for the  solid lines approach at a 2056^2 resolution.

- **Solid lines**


- **Dashes lines**

**Comment by luraess** [comment]
Thanks for the coo results. Would need to check further, but looks that the **solid lines** provide cleaner results which are more consistent with lower resolution ones. Seems in the **dashed line** approach, regularisation is somehow different at higher res.

**Comment by albert-de-montserrat** [comment]
Alright, both approaches functional in 3D as well.

*Solid lines* (shear stress interpolated from centers to mid-edges) 
3D_reg_vep_centers

*Dashed lines* (viscosity interpolated from centers to mid-edges, then shear stress is computed) 
3D_reg_vep

**Comment by luraess** [comment]
3D looks cool! Are both example run at same resolution?

**Comment by albert-de-montserrat** [comment]
Not exactly. Bottom one is 128^3 while the top one is 102^3, for some reason my GPU was getting stuck at 128^3 today .

**Comment by luraess** [comment]
Solid line version seems somehow "smoother", thus my question.

**Comment by albert-de-montserrat** [comment]
No, you are right. I uploaded the wrong snapshot of the bottom one >.< i will edit the post

**Comment by luraess** [comment]
Yay - thanks!

**Pull request merged**